cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,333 @@
+	for _, f := range files {
+		fname := path.Join(ipAllocPath, f.Name())
+
+		// Delete and skip the last reserved IP.
+		if strings.Contains(f.Name(), "last") || f.Name() == "lock" {
+			if err = os.Remove(fname); err != nil {
+				return fmt.Errorf("failed to remove file %s: %s", fname, err)
+			}
+			continue
+		}
+
+		// The name of the file is its IP address.
+		ip, _, err := cnet.ParseCIDR(fmt.Sprintf("%s/32", f.Name()))
+		if err != nil {
+			return fmt.Errorf("failed to parse IP %s: %s", f.Name(), err)
+		}
+
+		// The contents are the container ID.
+		b, err := ioutil.ReadFile(fname)
+		if err != nil {
+			return fmt.Errorf("failed to read file %s: %s", fname, err)
+		}
+		containerID := string(b)
+
+		// See if the IP hasn't already been assigned for the handleID.
+		handleID := utils.GetHandleID("k8s-pod-network", containerID)
+		ips, err := c.IPAM().IPsByHandle(ctxt, handleID)
+		if err == nil {
+		} else if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			// Return error if it is NOT does not exist, otherwise continue.
+			return fmt.Errorf("failed to fetch ips for handleID %s: %s", handleID, err)
+		}
+
+		alreadyExists := false
+		for _, ip := range ips {
+			if _, ok := podIPMap[ip.String()]; ok {
+				alreadyExists = true
+				break
+			}
+		}
+		if alreadyExists {
+			continue
+		}
+
+		// Get the pod resource associated with the IP.
+		pod, ok := podIPMap[f.Name()]
+		if !ok {
+			log.WithField("ip", f.Name()).Info("pod not found for IP, deleting and continuing...")
+			if err = os.Remove(fname); err != nil {
+				return fmt.Errorf("failed to remove file %s: %s", fname, err)
+			}
+			continue
+		}
+
+		// Store allocation to datastore
+		if err = c.IPAM().AssignIP(ctxt, ipam.AssignIPArgs{
+			IP:       *ip,
+			HandleID: &handleID,
+			Hostname: nodename,
+			Attrs: map[string]string{
+				ipam.AttributeNode:      nodename,
+				ipam.AttributePod:       pod.Name,
+				ipam.AttributeNamespace: pod.Namespace,
+			},
+		}); err != nil {
+			return fmt.Errorf("failed to assign IP to calico backend: %s", err)
+		}
+
+		// Delete the file
+		if err = os.Remove(fname); err != nil {
+			return fmt.Errorf("failed to remove file %s: %s", fname, err)
+		}
+	}
+
+	// HACK: Always re-enable datastoreReady.
+	// TODO: Should check DaemonSet rolling update status to see if this is the last node to update and
+	// only re-enable datastoreReady if the number of UpdatedNumberScheduled is MaxUnavailable less than the DesiredNumberScheduled.
+	// calicoDS, err := k8sClient.AppsV1().DaemonSets("kube-system").Get("calico-node", metav1.GetOptions{})
+	// if (calicoDS.Status.UpdatedNumberScheduled == calicoDS.Status.DesiredNumberScheduled-calicoDS.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable.IntVal)
+	*clusterInfo.Spec.DatastoreReady = true
+	if _, err = c.ClusterInformation().Update(ctxt, clusterInfo, options.SetOptions{}); err != nil {
+		return fmt.Errorf("failed to re-enable cluster: %s", err)
+	}
+
+	return nil
```

Comment:
I think if we exit successfully we should remove the `.disabled` binaries so they don't sit around forever.

Reply:
agreed